internal/venice-common/src/main/java/com/linkedin/venice/utils/ArrayCollection.java

```diff
@@ -0,0 +1,147 @@
+package com.linkedin.venice.utils;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.function.IntSupplier;
+
+
+/**
+ * Expose an array of {@link E} as an immutable {@link java.util.Collection<E>}
+ *
+ * Changes to the array are reflected in the collection.
+ */
+public class ArrayCollection<E> implements Collection<E> {
+  private final E[] array;
+  private final IntSupplier populatedSizeSupplier;
+
+  public ArrayCollection(E[] array) {
+    this.array = array;
+    this.populatedSizeSupplier = () -> {
+      int populatedSize = 0;
+      for (E item: array) {
+        if (item != null) {
+          populatedSize++;
+        }
+      }
+      return populatedSize;
+    };
+  }
+
+  /**
+   * @param array backing the collection
+   * @param populatedSizeSupplier for when the caller has a more efficient way of knowing the populated size of the array
+   */
+  public ArrayCollection(E[] array, IntSupplier populatedSizeSupplier) {
```

Comment:
As a general utility, shall we check whether `IntSupplier` will produce a size, which is larger than the underlying `array` length?

Reply:
Sure, I can add that.